src/librustc_builtin_macros/format.rs

```diff
@@ -135,7 +136,43 @@ fn parse_args<'a>(
         return Err(ecx.struct_span_err(sp, "requires at least a format string argument"));
     }
 
-    let fmtstr = p.parse_expr()?;
+    let first_token = &p.token;
+    let fmtstr = match first_token.kind {
+        token::TokenKind::Literal(lit_token @ token::Lit { kind: token::LitKind::Str, .. }) => {
+            // If the first token is a string literal, then a format expression
+            // is constructed from it and the parser is bumped, so that it
+            // points to the next token.
+            //
+            // This allows us to properly handle cases when the first comma
+            // after the format string is mistakenly replaced with any operator,
+            // which cause the expression parser to eat too much tokens.
+            let span = first_token.span;
+            let lit = ast::Lit::from_token(&first_token).unwrap_or_else(|_| {
+                // If the format string can not be represented as a string
+                // literal (most likely because it contains escaping error),
+                // then a fake one is generated.
+                let mut lit_token = lit_token;
+                lit_token.kind = token::LitKind::Err;
+
+                ast::Lit { token: lit_token, kind: ast::LitKind::Err(lit_token.symbol), span }
+            });
+
+            let kind = ast::ExprKind::Lit(lit);
+
+            let attrs = ThinVec::new();
+            let tokens = None;
+            let id = rustc_ast::DUMMY_NODE_ID;
+
+            p.bump();
+
+            P(ast::Expr { span, kind, attrs, tokens, id })
```

Comment:
You could use `p.parse_literal_maybe_minus()` to avoid implementing this logic by hand. (It's not currently public but it's ok to make it public.)